### DISCUSSION TITLE
Best Practice Question: IPC using external component e.g. captive portal system service

### DISCUSSION THREAD

**Discussion by MartinLoeper** [aws4embeddedlinux/meta-aws#1090] [Q&A] [answered]
Hi, I have a question regarding IPC on our Greengrass v2 device which is built using yocto.

We have a captive portal for initial device configuration (networking, provision by trusted user flow, etc.).
The captive portal is not deployed as a greengrass component.

The [greengrass v2 documentation regarding IPC](https://docs.aws.amazon.com/greengrass/v2/developerguide/interprocess-communication.html) sounds like IPC is only possible among components running inside the greengrass runtime. Is that interpretation correct?

If yes, is there an option to query deployment status from an external system service such as a captive portal?
I know that there is the greengrass-cli which is capable to do that. Is it possible to interface with the nucleus core from our captive portal just like the greengrass-cli binary does?

**Comment by nenadilic84** [comment]
Hi Martin,

Thanks for posing this questions. I'm actually thinking on doing a small blog post on how to interface GGv2, stay tuned for that.

As to answer your question you would need 2 variables in order to access the IPC:
1. `AWS_GG_NUCLEUS_DOMAIN_SOCKET_FILEPATH_FOR_COMPONENT ` which provides that path to the IPC socket
2. `SVCUID ` which is a secret token that components use to the IPC, which is frequently refreshed.

If there is no option for you to have the captive portal deployed as a GGv2 component, I suggest you:
1. either create a proxy component that exposes functionalities needed by the captive portal and use standard methods to interface it, or
2. simply call the `greengrass-cli` to get from the captive portal, but please note that the is deployed as a GGv2 component so you would need to do that first.


More about env variables available inside the greengrass component: https://docs.aws.amazon.com/greengrass/v2/developerguide/component-environment-variables.html

**Comment by MartinLoeper** [comment]
Hi @nenadilic84! Thanks for the thorough and timely response!

The second option sounds promising! Do you know whether it is possible to provision components in advance using yocto?
More precisely: Is it possible to bake an initial deployment into the yocto image which ships e.g. the greengrass-cli?

**Comment by nenadilic84** [comment, marked as answer]
Yes, that would definitely be possible.
To do this you would need to append files to the original greengrass recipe that adds the `greengrass-cli` component.

On which files you would need to add and their location I suggest you take a look at the installation after using the easy setup on, for example on ubuntu:

```
sudo -E java -Droot="/greengrass/v2" -Dlog.store=FILE \
  -jar ./GreengrassInstaller/lib/Greengrass.jar \
  --aws-region region \
  --thing-name MyGreengrassCore \
  --thing-group-name MyGreengrassCoreGroup \
  --thing-policy-name GreengrassV2IoTThingPolicy \
  --tes-role-name GreengrassV2TokenExchangeRole \
  --tes-role-alias-name GreengrassCoreTokenExchangeRoleAlias \
  --component-default-user ggc_user:ggc_group \
  --provision true \
  --setup-system-service true \
  --deploy-dev-tools true
```

Where you have the `  --deploy-dev-tools true` flag set.

This will provide the layout for packages directory that includes the `greengrass-cli`.

I tried this long time ago and it did work, but if you are doing this, please let me know as we might want to include this in the recipe as well.

**Reply by MartinLoeper** [reply]
Thanks, will try that! :)

**Reply by MartinLoeper** [reply]
Do you think I can just copy the greengrass-cli binary and its dependencies and bake it into the image?
I will try that tomorrow. Thanks!! :)
